lib/util/list_files.js

```diff
@@ -0,0 +1,109 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert  = require("chai").assert;
+const co      = require("co");
+const NodeGit = require("nodegit");
+const path    = require("path");
+
+const SubmoduleConfigUtil = require("./submodule_config_util");
+const SubmoduleUtil       = require("./submodule_util");
+
+/**
+ * Return the files in the index of the specified `repo`.
+ *
+ * @async
+ * @param {NodeGit.Repository} repo
+ * @return [String]
+ */
+const listFilesForRepo = co.wrap(function *(repo) {
+    const index = yield repo.index();
+    const entries = index.entries();
+    return entries.map(entry => entry.path);
+});
+
+/**
+ * Return a list the files in the specified `repo` and its *open* submodules,
+ * relative to the root of the repository, or to the optionally specified
+ * `relativePath` if provided.  Note that the order of the list is
+ * not defined.  Note that if `relativePath` is provided, no files will be
+ * listed that are not within that subdirectory.  The behavior is undefined
+ * unless `relativePath` (if provided) is a valid relative path.
+ *
+ * @async
+ * @param {NodeGit.Repository} repo
+ * @param {String}             [relativePath]
+ * @return [String]
+ */
+exports.listFiles = co.wrap(function *(repo, relativePath) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    if (undefined !== relativePath && null !== relativePath) {
+        assert.isString(relativePath);
+        assert.notEqual(relativePath, "");
+        assert(!path.isAbsolute(relativePath));
+    }
+    const subNames = new Set(yield SubmoduleUtil.getSubmoduleNames(repo));
+    let result = (yield listFilesForRepo(repo)).filter(name => {
+        return !subNames.has(name) &&
+            SubmoduleConfigUtil.modulesFileName !== name;
+    });
+    const openSubs = yield SubmoduleUtil.listOpenSubmodules(repo);
+    const subFiles = yield openSubs.map(co.wrap(function *(name) {
+
+        // If we have a relative path, optimize by skipping submodules whose
+        // contents would be omitted.  Include the contents of this submodule
+        // if the relative path is one of the roots of the submodules path (its
+        // name starts with relative path), or if the relative path is actually
+        // inside the submodule (the submodule name is a prefix of the relative
+        // path).
+
+        if (relativePath &&
+           !(name.startsWith(relativePath) || relativePath.startsWith(name))) {
```

Comment:
I'm worried about name="foobar/baz" and relativePath="foo" here -- or does relativePath always have a trailing slash? If so, the preconditions should check this.